service/history/shardContext.go

```diff
@@ -242,6 +244,13 @@ func (s *shardContextImpl) UpdateDomainNotificationVersion(domainNotificationVer
 	return s.updateShardInfoLocked()
 }
 
+func (s *shardContextImpl) UpdateActiveTimerReadLevel(readLevel time.Time) {
+	s.Lock()
```

Comment:
This might has performance impact on timerqueue read tasks, but if changed with property lock, there still might be race condition which requires more changes. I am looking into other possible solution